dynamic-check/src/objects/frame.rs

```diff
@@ -0,0 +1,1198 @@
+    ///
+    fn parse_negate_expression(&mut self, expression: &Expression, span: &Span) -> Result<Type, FrameError> {
+        // Parse the expression type.
+        let type_ = self.parse_expression(expression)?;
+
+        // Assert that this integer can be negated.
+        self.assert_negative_integer(&type_, span);
+
+        Ok(type_)
+    }
+
+    ///
+    /// Returns the `Boolean` type if the binary expression is a `Boolean` type.
+    ///
+    fn parse_boolean_binary_expression(
+        &mut self,
+        left: &Expression,
+        right: &Expression,
+        span: &Span,
+    ) -> Result<Type, FrameError> {
+        // Create the `Boolean` type.
+        let boolean_type = Type::Boolean;
+
+        // Create a new type assertion for the binary expression
+        let _binary_expression_type = self.parse_binary_expression(left, right, span)?;
+
+        // Return the `Boolean` type.
+        Ok(boolean_type)
+    }
+
+    ///
+    /// Returns the type of the conditional expression.
+    ///
+    fn parse_conditional_expression(
+        &mut self,
+        condition: &Expression,
+        first: &Expression,
+        second: &Expression,
+        span: &Span,
+    ) -> Result<Type, FrameError> {
+        // Check that the type of the condition expression is a boolean.
+        let _condition_type = self.parse_boolean_expression(condition, span)?;
+
+        // Check that the types of the first and second expression are equal.
+        self.parse_binary_expression(first, second, span)
+    }
+
+    ///
+    /// Returns the type of the tuple expression.
+    ///
+    fn parse_tuple(&mut self, expressions: &Vec<Expression>, _span: &Span) -> Result<Type, FrameError> {
+        let mut types = vec![];
+
+        // Parse all tuple expressions.
+        for expression in expressions {
+            let type_ = self.parse_expression(expression)?;
+
+            types.push(type_)
+        }
+
+        Ok(Type::Tuple(types))
+    }
+
+    ///
+    /// Returns the type of the accessed tuple element when called as an expression.
+    ///
+    fn parse_expression_tuple_access(
+        &mut self,
+        expression: &Expression,
+        index: usize,
+        span: &Span,
+    ) -> Result<Type, FrameError> {
+        // Parse the tuple expression which could be a variable with type tuple.
+        let type_ = self.parse_expression(expression)?;
+
+        // Parse the tuple access.
+        self.parse_tuple_access(type_, index, span)
+    }
+
+    ///
+    /// Returns the type of the accessed tuple element.
+    ///
+    fn parse_tuple_access(&mut self, type_: Type, index: usize, span: &Span) -> Result<Type, FrameError> {
+        // Check the type is a tuple.
+        let elements = match type_ {
+            Type::Tuple(elements) => elements,
+            type_ => return Err(FrameError::tuple_access(&type_, span)),
+        };
+
+        let element_type = elements[index].clone();
```

Comment:
since `type_` is consumed here, this can instead be:
```suggestion
        let element_type = elements.swap_remove(index);
```
which is `O(1)` and removes the need for the `clone()`; the `let elements` assignment above would only need to be `mut`